### PR TITLE
feat:The limit-conn plugin supports 'consumer_name' field

### PR DESCRIPTION
**Pull request by Jaycean**
### What this PR does / why we need it:
- The limit-conn plugin supports consumer_name field
- The method of PR is referred: #2333  @Firstsawyou

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible?

### CODE REVIEW

File: t/plugin/limit-conn.t [hunk 1]
```diff
@@ -956,3 +956,45 @@ GET /test_concurrency
...
+            end
+            ngx.say(body)
+        }
+    }
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
```
Comment (liuxiran) [inline t/plugin/limit-conn.t:1000]: It would be better to add more test case

**Review by liuxiran — CHANGES_REQUESTED** [review]

File: t/plugin/limit-conn.t (hunk 1 above)
Comment (Jaycean) [reply t/plugin/limit-conn.t]: done.

**Review by liuxiran — APPROVED** [review]

File: doc/zh-cn/plugins/limit-conn.md [hunk 2]
```diff
@@ -30,7 +30,7 @@
...
-| key                | object  | required |        | ["remote_addr", "server_addr", "http_x_real_ip", "http_x_forwarded_for"] | 用户指定的限制并发级别的关键字，可以是客户端IP或服务端IP。<br>例如，可以使用主机名(或服务器区域)作为关键字，以便限制每个主机名的并发性。 否则，我们也可以使用客户端地址作为关键字，这样我们就可以避免单个客户端用太多的并行连接或请求淹没我们的服务。 <br>当前接受的 key 有："remote_addr"(客户端IP地址), "server_addr"(服务端 IP 地址), 请求头中的"X-Forwarded-For" 或 "X-Real-IP"。 |
+| key                | object  | required |        | ["remote_addr", "server_addr", "http_x_real_ip", "http_x_forwarded_for", "consumer_name"] | 用户指定的限制并发级别的关键字，可以是客户端IP或服务端IP。<br>例如，可以使用主机名(或服务器区域)作为关键字，以便限制每个主机名的并发性。 否则，我们也可以使用客户端地址作为关键字，这样我们就可以避免单个客户端用太多的并行连接或请求淹没我们的服务。 <br>当前接受的 key 有："remote_addr"(客户端IP地址), "server_addr"(服务端 IP 地址), 请求头中的"X-Forwarded-For" 或 "X-Real-IP", "consumer_name"(consumer 的 username)。 |
```
Comment (juzhiyuan) [inline doc/zh-cn/plugins/limit-conn.md:33]: Please use `（` in Chinese.

File: t/plugin/limit-conn.t [hunk 3]
```diff
@@ -956,3 +956,216 @@ GET /test_concurrency
...
+
+
+
+=== TEST 26: create route and consumer_name is consumer_jack
```
[the remark this answers is not in the document]
Comment (Jaycean) [reply t/plugin/limit-conn.t:993]: consumer_ Jack is an example value created by consumer_name

[the remark this answers is not in the document]
Comment (Jaycean) [reply t/plugin/limit-conn.t:993]: consumer_name:consumer_ Jack->key:value

File: doc/zh-cn/plugins/limit-conn.md (hunk 2 above)
Comment (Jaycean) [reply doc/zh-cn/plugins/limit-conn.md]: done.

File: t/plugin/limit-conn.t (hunk 3 above)
[the remark this answers is not in the document]
Comment (juzhiyuan) [reply t/plugin/limit-conn.t:993]: Got it

**Comment by Jaycean** [comment]
@spacewander PTAL

File: t/plugin/limit-conn.t (hunk 3 above)
Comment (membphis) [inline t/plugin/limit-conn.t:993]: bad title.

it should be `create route and enable plugin 'key-auth'`

File: t/plugin/limit-conn.t [hunk 4]
```diff
@@ -956,3 +956,216 @@ GET /test_concurrency
...
+200
+200
+200
+200
+--- no_error_log
+[error]
+
+
+
+=== TEST 28: update plugin
```
Comment (membphis) [inline t/plugin/limit-conn.t:1084]: bad title too

Comment (membphis) [reply t/plugin/limit-conn.t:1084]: need to write more description

**Comment by membphis** [comment]
we need to print the value of the `limit key`, and check it in error log.

**Comment by Jaycean** [comment]
> we need to print the value of the `limit key`, and check it in error log.

I'm sorry that I didn't understand what you meant. I added the test error value of the key and the test case of Check Schema, so I don't know if I understood them correctly

**Comment by membphis** [comment]
@Jaycean 

https://github.com/apache/apisix/blob/d63f3e068c0d54ab7b2b3acec8e5f4914da0a4b7/apisix/plugins/limit-conn.lua#L71

we can print the `ctx.var[conf.key]` to error log before this line, and confirm it in the test case.

test case example: https://github.com/apache/apisix/blob/d63f3e068c0d54ab7b2b3acec8e5f4914da0a4b7/t/plugin/limit-conn.t#L705-L706

**Comment by Jaycean** [comment]
> @Jaycean
> 
> https://github.com/apache/apisix/blob/d63f3e068c0d54ab7b2b3acec8e5f4914da0a4b7/apisix/plugins/limit-conn.lua#L71
> 
> we can print the `ctx.var[conf.key]` to error log before this line, and confirm it in the test case.
> 
> test case example:
> 
> https://github.com/apache/apisix/blob/d63f3e068c0d54ab7b2b3acec8e5f4914da0a4b7/t/plugin/limit-conn.t#L705-L706

Thank you for your patient explanation

File: t/plugin/limit-conn.t (hunk 4 above)
Comment (Jaycean) [reply t/plugin/limit-conn.t]: done

File: t/plugin/limit-conn.t (hunk 3 above)
Comment (Jaycean) [reply t/plugin/limit-conn.t]: done

**Comment by Jaycean** [comment]
> @Jaycean
> 
> https://github.com/apache/apisix/blob/d63f3e068c0d54ab7b2b3acec8e5f4914da0a4b7/apisix/plugins/limit-conn.lua#L71
> 
> we can print the `ctx.var[conf.key]` to error log before this line, and confirm it in the test case.
> 
> test case example:
> 
> https://github.com/apache/apisix/blob/d63f3e068c0d54ab7b2b3acec8e5f4914da0a4b7/t/plugin/limit-conn.t#L705-L706

done.

**Review by membphis — APPROVED** [review]

**Comment by membphis** [comment]
@Jaycean please rebase your master

image

**Comment by Jaycean** [comment]
@membphis Done. PTAL

**Comment by membphis** [comment]
@Jaycean I think we can add a checkpoint for confirming if we fetched a correct `consumer_name`.

https://github.com/apache/apisix/blob/49818548f5510a0789c93519aaedd08c35ae69f8/apisix/plugins/limit-conn.lua#L72

**Comment by Jaycean** [comment]
> @Jaycean I think we can add a checkpoint for confirming if we fetched a correct `consumer_name`.
> 
> https://github.com/apache/apisix/blob/49818548f5510a0789c93519aaedd08c35ae69f8/apisix/plugins/limit-conn.lua#L72

There may be new fields added to the `consumer_name` field in the future, the default is `""`, to determine whether the current field is in the optional range? If there are new fields, continue to add? Because I think that if the wrong field is selected, the plugin is actually unavailable.

**Comment by membphis** [comment]
```
--- error_log
limit key: consumer_jackroute
```

to

```
--- error_log_like eval
qr/limit key: consumer_jackroute&consumer\d+/
```

**Review by membphis — APPROVED** [review]
LGTM, except minor things in the test cases.

File: t/plugin/limit-conn.t [hunk 5]
```diff
@@ -986,3 +985,244 @@ property "default_conn_delay" validation failed: expected 0 to be sctrictly grea
...
+        local reqs = {}
+        for i = 1, 10 do
+            reqs[i] = { "/access_root_dir" }
+        end
+        local resps = { ngx.location.capture_multi(reqs) }
+        for i, resp in ipairs(resps) do
+            ngx.say(resp.status)
+        end
+    }
+}
```
Comment (Firstsawyou) [inline t/plugin/limit-conn.t:1094]: I think 4 spaces need to be indented.

File: t/plugin/limit-conn.t [hunk 6]
```diff
@@ -986,3 +985,244 @@ property "default_conn_delay" validation failed: expected 0 to be sctrictly grea
...
```
Comment (Firstsawyou) [inline t/plugin/limit-conn.t:1185]: Need to indent 4 spaces.

**Review by Firstsawyou — CHANGES_REQUESTED** [review]

File: t/plugin/limit-conn.t (hunk 5 above)
Comment (Jaycean) [reply t/plugin/limit-conn.t]: Yes, there are indeed some problems with the format, thank you for your suggestions

Comment (Jaycean) [reply t/plugin/limit-conn.t]: done

File: t/plugin/limit-conn.t (hunk 6 above)
Comment (Jaycean) [reply t/plugin/limit-conn.t]: done

**Comment by Jaycean** [comment]
> LGTM, except minor things in the test cases.

done.

**Comment by membphis** [comment]
@Jaycean many thx, merged

**Pull request merged**